src/python/pants/engine/scheduler.py

```diff
@@ -71,23 +70,44 @@ class ExecutionError(Exception):
   pass
 
 
-class WrappedNativeScheduler(object):
-  def __init__(self, native, build_root, work_dir, ignore_patterns, rule_index):
+class Scheduler(object):
+  def __init__(self, native, project_tree, work_dir, rules, include_trace_on_error=True, validate=True):
+    """
+    :param native: An instance of engine.native.Native.
+    :param project_tree: An instance of ProjectTree for the current build root.
+    :param work_dir: The pants work dir.
+    :param rules: A set of Rules which is used to compute values in the graph.
+    :param include_trace_on_error: Include the trace through the graph upon encountering errors.
+    :type include_trace_on_error: bool
+    :param validate: True to assert that the ruleset is valid.
+    """
     self._native = native
+    self.include_trace_on_error = include_trace_on_error
+
     # TODO: The only (?) case where we use inheritance rather than exact type unions.
     has_products_constraint = SubclassesOf(HasProducts)
+
+    # Validate and register all provided and intrinsic tasks.
+    rule_index = RuleIndex.create(list(rules))
     self._root_subject_types = sorted(rule_index.roots)
 
-    # Create the ExternContext, and the native Scheduler.
+    # If configured, visualize the rule graph before asserting that it is valid.
+    if self.visualize_to_dir() is not None:
+      rule_graph_name = 'rule_graph.dot'
+      self.visualize_rule_graph_to_file(os.path.join(self.visualize_to_dir(), rule_graph_name))
+
+    # Create the native Scheduler and Session.
+    # TODO: This `_tasks` reference could be a local variable, since it is not used
```

Comment:
Any reason not to now?

Reply:
I was trying not to get sucked into sideprojects, since I generally make PRs that are way too large. But sure, why not!

Reply:
...and actually, I think it might be a non-trivial sideproject. Going to leave it as is for now.